pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,168 @@
+package restrictusers
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+)
+
+func init() {
+	admission.RegisterPlugin("RestrictSubjectBindings",
+		func(kclient kclientset.Interface, config io.Reader) (admission.Interface,
+			error) {
+			return NewRestrictUsersAdmission(kclient)
+		})
+}
+
+// restrictUsersAdmission implements admission.Interface and enforces
+// restrictions on adding rolebindings in a project to permit only designated
+// subjects.
+type restrictUsersAdmission struct {
+	*admission.Handler
+
+	oclient oclient.Interface
+	kclient kclientset.Interface
+}
+
+var _ = oadmission.WantsOpenshiftClient(&restrictUsersAdmission{})
+var _ = oadmission.Validator(&restrictUsersAdmission{})
+
+// NewRestrictUsersAdmission configures an admission plugin that enforces
+// restrictions on adding role bindings in a project.
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		kclient: kclient,
+	}, nil
+}
+
+func (q *restrictUsersAdmission) SetOpenshiftClient(c oclient.Interface) {
+	q.oclient = c
+}
+
+// rolebindingsDelta returns the relative complement of rolebindings a in
+// rolebindings b (i.e., a∖b).
+func rolebindingsDelta(a, b *authorizationapi.RoleBinding) []kapi.ObjectReference {
+	result := []kapi.ObjectReference{}
+
+	for _, y := range b.Subjects {
+		for _, x := range a.Subjects {
+			if y == x {
+				goto nexty
+			}
+		}
+		result = append(result, y)
+	nexty:
+	}
+
+	return result
+}
+
+// Admit makes admission decisions that enforce restrictions on adding
+// project-scoped role-bindings.  In order for a role binding to be permitted,
+// each subject in the binding must be matched by some rolebinding restriction
+// in the namespace.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if a.GetResource().GroupResource() != authorizationapi.Resource("rolebindings") {
```

Comment:
you need to catch policybindings too.  You should coerce the policybindings down to the different rolebindings and check them each individually.